### ISSUE TITLE
Add Resume/CV analysis (its fine if its simple)

### ISSUE THREAD

**Issue by g-charan** [g-charan/Job_auto#5]
File name.pdf

1. Load the pdf file
2. Optional: embeddings -
3. Give the loaded read (Extracte Values) values to LLM in prompt manner
   "This is the data  $values, with this context provide the summary in this format
   gmail -
   name- 
   phone number-
   "
4. Response from LLM, **OUTPUT**

**Comment by Neeraj-nk3** [comment]
# ======================================================================
#  ENHANCED RESUME ANALYZER – PDF / DOCX – GOOGLE GEMINI API
#  (OCR fallback, skill normalization, soft-skills extraction, confidence
#   scores, email/phone validation, plain text output for GitHub terminal)
# ======================================================================

import os
import re
import json
from typing import Dict, Any, List, Optional

from pypdf import PdfReader
import docx2txt
import fitz               # PyMuPDF  (pip install pymupdf)
import pytesseract         # OCR for scanned PDFs
from PIL import Image
from google import genai   # pip install google-genai
from dotenv import load_dotenv


# ---------------------  LOW-LEVEL UTILITIES  --------------------------

def extract_text_from_pdf(path: str) -> str:
    """Extract text; OCR fallback for image-based pages."""
    text = ""
    try:
        reader = PdfReader(path)
        for page in reader.pages:
            page_text = page.extract_text() or ""
            text += "\n" + page_text
    except Exception:
        pass  # continue to OCR fallback

    if text.strip():  # good native text found
        return text.strip()

    # ---------- OCR fallback ----------
    doc = fitz.open(path)
    for page in doc:
        pix = page.get_pixmap()
        img = Image.frombytes("RGB", [pix.width, pix.height], pix.samples)
        text += "\n" + pytesseract.image_to_string(img)
    return text.strip()


def extract_text_from_docx(path: str) -> str:
    return docx2txt.process(path).strip()


EMAIL_RE = re.compile(r"^[\w\.-]+@[\w\.-]+\.\w+$")
PHONE_RE = re.compile(r"^\+?[\d\s\-\(\)]{7,}$")


def is_valid_email(email: str) -> bool:
    return bool(EMAIL_RE.fullmatch(email.strip()))


def is_valid_phone(phone: str) -> bool:
    return bool(PHONE_RE.fullmatch(phone.strip()))


def normalize_skills(skills: List[str]) -> List[str]:
    """Map synonyms → canonical forms & de-duplicate."""
    mapping = {
        "js": "JavaScript", "javascript": "JavaScript",
        "py": "Python", "react": "React.js",
        "node": "Node.js", "nodejs": "Node.js"
    }
    normalized = []
    for s in skills:
        key = s.strip().lower()
        normalized.append(mapping.get(key, s.strip()))
    return sorted(set(normalized))


# ---------------------  ANALYZER CLASS  -------------------------------

class EnhancedResumeAnalyzer:
    """Resume parser + Gemini summarizer with confidence scores."""
    def __init__(self, api_key: Optional[str] = None):
        load_dotenv()
        self.api_key = api_key or os.getenv("GOOGLE_API_KEY")
        if not self.api_key:
            raise ValueError("Google API key missing (set GOOGLE_API_KEY).")
        genai.configure(api_key=self.api_key)
        self.model = genai.GenerativeModel("gemini-1.5-flash")

    # ------------ file loading ------------
    def load_text(self, path: str) -> str:
        if not os.path.isfile(path):
            raise FileNotFoundError(path)
        ext = path.lower().split(".")[-1]
        if ext == "pdf":
            return extract_text_from_pdf(path)
        if ext in {"docx", "doc"}:
            return extract_text_from_docx(path)
        raise ValueError("Unsupported file format: " + ext)

    # ------------ prompt for Gemini ------------
    def build_prompt(self, text: str) -> str:
        return f"""
You are an expert resume extractor.
Here is raw resume text:
{text}

Return ONLY a JSON object with:
  basic_info  (name, email, phone, location)
  professional_summary
  technical_skills {{
       programming_languages, frameworks_libraries,
       tools_technologies, databases, cloud_platforms, other_skills
  }}
  soft_skills              (array)
  experience               (company, role, duration, key_responsibilities)
  education                (degree, institution, year, gpa)
  certifications           (array)
  suitable_roles           (array)
  confidence_scores        (0-1 float for basic_info, skills, experience, edu)

If a value is unknown write "Not specified".
Return STRICT VALID JSON — no code-fences.
"""

    # ------------ Gemini call ------------
    def call_gemini(self, text: str) -> Dict[str, Any]:
        prompt = self.build_prompt(text)
        resp = self.model.generate_content(prompt)
        try:
            return json.loads(resp.text.strip())
        except Exception as e:
            raise ValueError("Gemini response invalid JSON") from e

    # ------------ post-processing ------------
    def clean(self, data: Dict[str, Any]) -> Dict[str, Any]:
        basic = data.get("basic_info", {})
        if "email" in basic:
            basic["email_valid"] = bool(is_valid_email(basic["email"]))
        if "phone" in basic:
            basic["phone_valid"] = bool(is_valid_phone(basic["phone"]))

        tech = data.get("technical_skills", {})
        for cat, skills in tech.items():
            tech[cat] = normalize_skills(skills)

        return data

    # ------------ helper for plain text output ------------
    def dict_to_text(self, d: Dict[str, Any], indent: int = 0) -> List[str]:
        lines = []
        for k, v in d.items():
            prefix = "  " * indent + str(k) + ": "
            if isinstance(v, dict):
                lines.append(prefix)
                lines.extend(self.dict_to_text(v, indent + 1))
            elif isinstance(v, list):
                lines.append(prefix)
                for item in v:
                    if isinstance(item, dict):
                        lines.extend(self.dict_to_text(item, indent + 1))
                    else:
                        lines.append("  " * (indent + 1) + str(item))
            else:
                lines.append(prefix + str(v))
        return lines

    # ------------ public ------------
    def analyze(self, path: str) -> Dict[str, Any]:
        print(f"→ Loading: {path}")
        text = self.load_text(path)
        print(f"  Extracted {len(text)} characters")
        raw = self.call_gemini(text)
        final = self.clean(raw)
        return final

    def analyze_and_print_text(self, path: str):
        """Analyze resume and print output in readable plain text."""
        result = self.analyze(path)
        print("\n==========  PARSED RESUME OUTPUT (TEXT)  ==========")
        lines = self.dict_to_text(result)
        print("\n".join(lines))


# ---------------------  QUICK DEMO  -----------------------------------

if __name__ == "__main__":
    analyzer = EnhancedResumeAnalyzer(api_key="YOUR_GOOGLE_API_KEY")
    analyzer.analyze_and_print_text("sample_resume.pdf")  # Change to your resume path

**Comment by Neeraj-nk3** [comment]
import os
import re
import json
from typing import Dict, Any, List, Optional

from pypdf import PdfReader
import docx2txt
import fitz
import pytesseract
from PIL import Image
from google import genai
from dotenv import load_dotenv

def extract_text_from_pdf(path: str) -> str:
    text = ""
    try:
        reader = PdfReader(path)
        for page in reader.pages:
            page_text = page.extract_text() or ""
            text += "\n" + page_text
    except Exception:
        pass
    if text.strip():
        return text.strip()
    doc = fitz.open(path)
    for page in doc:
        pix = page.get_pixmap()
        img = Image.frombytes("RGB", [pix.width, pix.height], pix.samples)
        text += "\n" + pytesseract.image_to_string(img)
    return text.strip()

def extract_text_from_docx(path: str) -> str:
    return docx2txt.process(path).strip()

EMAIL_RE = re.compile(r"^[\w\.-]+@[\w\.-]+\.\w+$")
PHONE_RE = re.compile(r"^\+?[\d\s\-\(\)]{7,}$")

def is_valid_email(email: str) -> bool:
    return bool(EMAIL_RE.fullmatch(email.strip()))

def is_valid_phone(phone: str) -> bool:
    return bool(PHONE_RE.fullmatch(phone.strip()))

def normalize_skills(skills: List[str]) -> List[str]:
    mapping = {
        "js": "JavaScript", "javascript": "JavaScript",
        "py": "Python", "react": "React.js",
        "node": "Node.js", "nodejs": "Node.js"
    }
    normalized = []
    for s in skills:
        key = s.strip().lower()
        normalized.append(mapping.get(key, s.strip()))
    return sorted(set(normalized))

class EnhancedResumeAnalyzer:
    def __init__(self, api_key: Optional[str] = None):
        load_dotenv()
        self.api_key = api_key or os.getenv("GOOGLE_API_KEY")
        if not self.api_key:
            raise ValueError("Google API key missing (set GOOGLE_API_KEY).")
        genai.configure(api_key=self.api_key)
        self.model = genai.GenerativeModel("gemini-1.5-flash")

    def load_text(self, path: str) -> str:
        if not os.path.isfile(path):
            raise FileNotFoundError(path)
        ext = path.lower().split(".")[-1]
        if ext == "pdf":
            return extract_text_from_pdf(path)
        if ext in {"docx", "doc"}:
            return extract_text_from_docx(path)
        raise ValueError("Unsupported file format: " + ext)

    def build_prompt(self, text: str) -> str:
        return f"""
You are an expert resume extractor.
Here is raw resume text:
{text}

Return ONLY a JSON object with:
  basic_info  (name, email, phone, location)
  professional_summary
  technical_skills {{
       programming_languages, frameworks_libraries,
       tools_technologies, databases, cloud_platforms, other_skills
  }}
  soft_skills              (array)
  experience               (company, role, duration, key_responsibilities)
  education                (degree, institution, year, gpa)
  certifications           (array)
  suitable_roles           (array)
  confidence_scores        (0-1 float for basic_info, skills, experience, edu)

If a value is unknown write "Not specified".
Return STRICT VALID JSON — no code-fences.
"""

    def call_gemini(self, text: str) -> Dict[str, Any]:
        prompt = self.build_prompt(text)
        resp = self.model.generate_content(prompt)
        try:
            return json.loads(resp.text.strip())
        except Exception as e:
            raise ValueError("Gemini response invalid JSON") from e

    def clean(self, data: Dict[str, Any]) -> Dict[str, Any]:
        basic = data.get("basic_info", {})
        if "email" in basic:
            basic["email_valid"] = bool(is_valid_email(basic["email"]))
        if "phone" in basic:
            basic["phone_valid"] = bool(is_valid_phone(basic["phone"]))
        tech = data.get("technical_skills", {})
        for cat, skills in tech.items():
            tech[cat] = normalize_skills(skills)
        return data

    def dict_to_text(self, d: Dict[str, Any], indent: int = 0) -> List[str]:
        lines = []
        for k, v in d.items():
            prefix = "  " * indent + str(k) + ": "
            if isinstance(v, dict):
                lines.append(prefix)
                lines.extend(self.dict_to_text(v, indent + 1))
            elif isinstance(v, list):
                lines.append(prefix)
                for item in v:
                    if isinstance(item, dict):
                        lines.extend(self.dict_to_text(item, indent + 1))
                    else:
                        lines.append("  " * (indent + 1) + str(item))
            else:
                lines.append(prefix + str(v))
        return lines

    def analyze(self, path: str) -> Dict[str, Any]:
        text = self.load_text(path)
        raw = self.call_gemini(text)
        final = self.clean(raw)
        return final

    def analyze_and_print_text(self, path: str):
        result = self.analyze(path)
        print("\n==========  PARSED RESUME OUTPUT (TEXT)  ==========")
        lines = self.dict_to_text(result)
        print("\n".join(lines))

if __name__ == "__main__":
    analyzer = EnhancedResumeAnalyzer(api_key="YOUR_GOOGLE_API_KEY")
    analyzer.analyze_and_print_text("sample_resume.pdf")

**Comment by Neeraj-nk3** [comment]
EXPECTED OUTPUT 
==========  PARSED RESUME OUTPUT (TEXT)  ==========

basic_info: 
  name: Rajesh Kumar Sharma
  email: rajesh.sharma@email.com
  email_valid: True
  phone: +91-9876543210
  phone_valid: True
  location: Bangalore, Karnataka, India

professional_summary: Mid-level software developer with 3+ years of full-stack experience in Python, JavaScript, and cloud deployments.

technical_skills: 
  programming_languages: 
    Python
    JavaScript
    Java
    C++
    SQL
  frameworks_libraries: 
    React.js
    Django
    Flask
    FastAPI
    Node.js
    Express.js
    Vue.js
  tools_technologies: 
    VS Code
    PyCharm
    Postman
    JIRA
    Git
    Docker
    Jenkins
  databases: 
    PostgreSQL
    MySQL
    MongoDB
    Redis
  cloud_platforms: 
    AWS
    Google Cloud Platform
  other_skills: 

soft_skills: 
  Communication
  Leadership
  Problem-solving
  Teamwork
  Adaptability

experience: 
  company: Tech Solutions Pvt Ltd
    role: Senior Software Developer
    duration: June 2022 - Present
    key_responsibilities: 
      Led a team of 3 developers
      Designed scalable backend APIs
      Optimized cloud deployment pipelines
  company: Digital Innovations Inc
    role: Software Developer
    duration: Jan 2021 - May 2022
    key_responsibilities: 
      Developed REST APIs
      Integrated third-party services
      Participated in agile sprints
  company: Startup Hub
    role: Software Developer Intern
    duration: Aug 2020 - Dec 2020
    key_responsibilities: 
      Assisted in building prototypes
      Worked on bug fixes
      Contributed to open-source tools

education: 
  degree: Bachelor of Technology in Computer Science
  institution: IIT Delhi
  year: 2020
  gpa: 8.6

certifications: 
  AWS Certified Developer
  Google Cloud Professional
  Python for Data Science
  React Developer

projects: 
  name: E-Commerce Platform
    description: Full-featured e-commerce site with payment integration.
    technologies: Python, Django, React.js
  name: Task Management System
    description: Web app for project and task tracking.
    technologies: FastAPI, Vue.js, MongoDB
  name: ML Stock Predictor
    description: Machine learning model for stock price prediction.
    technologies: Python, scikit-learn, TensorFlow

achievements: 
  Employee of the Month - March 2023
  Best Graduation Project Award

languages: 
  English
  Hindi

suitable_roles: 
  Senior Software Developer
  Full Stack Developer
  Backend Developer
  Python Developer
  React Developer
  Cloud Engineer

confidence_scores: 
  basic_info: 0.95
  skills: 0.90
  experience: 0.89
  education: 0.92

**Comment by Neeraj-nk3** [comment]
RESUME USED 
Rajesh Kumar Sharma
Senior Software Developer
Bangalore, Karnataka, India
Email: rajesh.sharma@email.com
Phone: +91-9876543210

Professional Summary:
Mid-level software developer with 3+ years of full-stack experience in Python, JavaScript, and cloud deployments. Skilled in building scalable web applications and leading small agile teams.

Technical Skills:
- Programming Languages: Python, JavaScript, Java, C++, SQL
- Frameworks/Libraries: React.js, Django, Flask, FastAPI, Node.js, Express.js, Vue.js
- Tools/Technologies: VS Code, PyCharm, Postman, JIRA, Git, Docker, Jenkins
- Databases: PostgreSQL, MySQL, MongoDB, Redis
- Cloud Platforms: AWS, Google Cloud Platform

Soft Skills:
Communication, Leadership, Problem-solving, Teamwork, Adaptability

Experience:
Tech Solutions Pvt Ltd
Senior Software Developer (June 2022 - Present)
- Led a team of 3 developers
- Designed scalable backend APIs
- Optimized cloud deployment pipelines

Digital Innovations Inc
Software Developer (Jan 2021 - May 2022)
- Developed REST APIs
- Integrated third-party services
- Participated in agile sprints

Startup Hub
Software Developer Intern (Aug 2020 - Dec 2020)
- Assisted in building prototypes
- Worked on bug fixes
- Contributed to open-source tools

Education:
Bachelor of Technology in Computer Science, IIT Delhi, 2020
GPA: 8.6

Certifications:
AWS Certified Developer
Google Cloud Professional
Python for Data Science
React Developer

Projects:
E-Commerce Platform
- Full-featured e-commerce site with payment integration.
- Technologies: Python, Django, React.js

Task Management System
- Web app for project and task tracking.
- Technologies: FastAPI, Vue.js, MongoDB

ML Stock Predictor
- Machine learning model for stock price prediction.
- Technologies: Python, scikit-learn, TensorFlow

Achievements:
Employee of the Month - March 2023
Best Graduation Project Award

Languages:
English, Hindi
